### PR TITLE
[CALCITE-3450] Support SetOp in RelMdTableReferences

### PR DESCRIPTION
**Pull request by xy2953396112**
Support Sample、SetOp、TableModify、Window in RelMdTableReferences.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdTableReferences.java [hunk 1]
```diff
@@ -183,6 +185,72 @@ protected RelMdTableReferences() {}
...
+        result.add(shiftTableRef);
+        currentTablesMapping.put(tableRef, shiftTableRef);
+      }
+      for (RelTableRef newRef : currentTablesMapping.values()) {
+        qualifiedNamesToRefs.put(newRef.getQualifiedName(), newRef);
+      }
+    }
+    return result;
+  }
+
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdTableReferences.java]: The implementation for getTableReferences of `Minus` and `Intersect` is basically the same with `Union`. Can you just put them together?

Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdTableReferences.java]: How about just to support SetOp and cover Union and Minus and Intersect.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 2]
```diff
@@ -2229,6 +2229,34 @@ private void checkAllPredicates(RelOptCluster cluster, RelOptTable empTable,
...
+    final RelNode rel2 = convertSql(sql2);
+    final RelMetadataQuery mq2 = rel2.getCluster().getMetadataQuery();
+    final Set<RelTableRef> tableReferences2 = Sets.newTreeSet(mq2.getTableReferences(rel2));
+    assertThat(tableReferences2.toString(),
+            equalTo("[[CATALOG, SALES, DEPT].#0, [CATALOG, SALES, EMP].#0]"));
+
+  }
+
+  @Test public void testTableReferenceForMinux() {
+    final String sql = "select emp.deptno, emp.sal from emp\n"
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java]: `testTableReferenceForMinus` ?

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdTableReferences.java (hunk 1 above)
Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdTableReferences.java]: Thanks,You're right.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java (hunk 2 above)
Comment (xy2953396112) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java]: changed it.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 3]
```diff
@@ -2229,6 +2229,34 @@ private void checkAllPredicates(RelOptCluster cluster, RelOptTable empTable,
...
+
+  @Test public void testTableReferenceForMinus() {
+    final String sql = "select emp.deptno, emp.sal from emp\n"
+            + "except all select emp.deptno, emp.sal from emp where empno = 5";
+    final RelNode rel = convertSql(sql);
+    final RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+    final Set<RelTableRef> tableReferences = Sets.newTreeSet(mq.getTableReferences(rel));
+    assertThat(tableReferences.toString(),
+            equalTo("[[CATALOG, SALES, EMP].#0, [CATALOG, SALES, EMP].#1]"));
+  }
```
Comment (yanlin-Lynn) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java]: why two cases in `testTableReferenceForIntersect`, but just one case in `testTableReferenceForMinus`?

Comment (xy2953396112) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java]: 'testTableReferenceForIntersect' is uesed for checking different tables.

**Pull request closed without merging**